src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,352 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal interface ITypeImportCompletionService : IWorkspaceService
+    {
+        ImmutableArray<TypeImportCompletionItem> GetAccessibleTopLevelTypesFromPEReference(
+            Solution solution,
+            Compilation compilation,
+            PortableExecutableReference peReference,
+            ImmutableHashSet<string> excludedNamespaces,
+            CancellationToken cancellationToken);
+
+        Task<ImmutableArray<TypeImportCompletionItem>> GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+            Solution solution,
+            Compilation compilation,
+            CompilationReference compilationReference,
+            ImmutableHashSet<string> excludedNamespaces,
+            CancellationToken cancellationToken);
+
+        Task<ImmutableArray<TypeImportCompletionItem>> GetAccessibleTopLevelTypesFromProjectAsync(
+            Project project,
+            ImmutableHashSet<string> excludedNamespaces,
+            CancellationToken cancellationToken);
+    }
+
+    [ExportWorkspaceServiceFactory(typeof(ITypeImportCompletionService), ServiceLayer.Editor), Shared]
+    internal sealed class TypeImportCompletionService : IWorkspaceServiceFactory
+    {
+        private readonly ConcurrentDictionary<string, ReferenceCacheEntry> _peItemsCache
+            = new ConcurrentDictionary<string, ReferenceCacheEntry>();
+
+        private readonly ConcurrentDictionary<ProjectId, ReferenceCacheEntry> _projectItemsCache
+            = new ConcurrentDictionary<ProjectId, ReferenceCacheEntry>();
```

Comment:
concurrent dictionaries can actually be more expensive than just using a dictionary+lock.  do we expect heavily concurrent access here?  or do we just want to have this stuff precomputed?

Reply:
No, I don't expect heavy concurrent here. Only case I can think of is the cache flush and cache read can happen at the same time. I will switch to lock+Dictionary (but either way, the impact should be insignificant)

Reply:
I'm about to add a solution crawler to flush invalid cache entries, not sure how much it will affect this. We will see